### DISCUSSION TITLE
best practice for create drived states form another atom

### DISCUSSION THREAD

**Discussion by dante01yoon** [pmndrs/jotai#1652] [Q&A]
Greetings, thanks for this excellent library!

If one Atom should be updated via another atom state  or normal variable, 
what would be best practice for achieving this ?

For instance , If there are  several search forms like categories, texts, dates and all of this forms are controlled via multiple atoms seperately.

Suppose  there is tab component, and  its ui  changes for any each form states changes.

I guess using custom hook for watching each form state change and update atom states related with tab component. 

```ts
// atoms for search forms 
const dateAtom = atom(new Date());
const textAtom  = atom("");

// atoms for ui component 

const tabType = {
 default: "default",
 ghost: "ghost",
}

const tabAtom = atom(tabType.default);

// custom hook
export const useTabAtom = () => {
  const [tabType, setTabType] = useAtom(tabAtom);
  const text = useAtomValue(textAtom);
  const prevTextRef = useRef(text);
  const { current: prevText } = prevTextRef;

  useMemo(() => {
    setTabType("default");
  }, [prevText])
}
```

I know using ref is kind of trivial ways to memoize previous states,
but might be better ways to use derived states  I guess ? 
--- 

Is using getter in atom doesn't cause atom states changes right? 

I guess below codes doesn't work? 
```ts

const previousText = "";
const tabAtom = atom(tabType.default);

const derivedTabAtom = atom(get => {
  const currentText =  get(textAtom); 
  if( previousText != currentText) {
     previousText = currentText
     return "default";
  }

 return get(tabAtom); 
});

```

**Comment by dante01yoon** [comment]
hmm.. I think the second approach is working?  this might be one approach? 
```ts 
let previousText = "";
const tabAtom = atom(tabType.default);

const derivedTabAtom = atom(get => {
  const currentText =  get(textAtom); 
  if( previousText != currentText) {
     previousText = currentText
     return "default";
  }

 return get(tabAtom); 
});

**Reply by dante01yoon** [reply]
I guess this is usual way, what a silly question 😅

**Reply by dai-shi** [reply]
No, `previousText` variable defined outside atoms is not good.

**Comment by dai-shi** [comment]
> what would be best practice for achieving this ?

The best would be defining atoms declaratively (in `read` function). It means we shouldn't depend on previous states.
Imperative code (like one depending on previous state) should only be in `write` function.

```js
const baseTextAtom = atom("");
export const tabAtom = atom("default");
export const textAtom = atom(
  (get) => get(baseTextAtom),
  (get, set, nextValue) => {
    const prevValue = get(baseTextAtom);
    set(baseTextAtom, nextValue);
    if (prevValue !== nextValue) {
      set(tabAtom, "default");
    }
  }
);
```

**Reply by dante01yoon** [reply]
But if there are multiple components' states depends on textAtom, textAtom should  know each states which is not good for seperation of concerns.  IMHO each seperate atoms should refer textAtom is better approach.

**Reply by dai-shi** [reply]
I think if you can make it independent from previous values, things can be declarative. Deriving states from another atom is basically the patterns jotai derived atoms can do, right?

**Reply by dante01yoon** [reply]
Maybe put action atom in the middle of two atoms  `source`, `destination` , give all the responsible to update `source` atoms could be another approach 

```ts
const sourceAtom = atom("");

const actionAtom = atom(null, (get, set, updater) => {
   const prevSourceValue = get(sourceAtom);
   set(sourceAtom, updater);
   if ( prevValue !== nextValue) {
        set( destinationAtom, "default")
   }
}); 
```

```tsx
const useSource = () => {
   const source = useAtomValue(sourceAtom);
   const update = useUpdateValue(actionAtom);

  return [source, update]; 
}
```
